feat(analytics): Analytics dashboard

oidacra opened the pull request:
## Summary

  This PR introduces a comprehensive analytics dashboard feature that provides data visualization and search capabilities for dotCMS. The implementation includes a
  complete data access layer, dashboard components, and enhanced user interface for analytics data exploration.

  ## Key Features

  - Complete Analytics Dashboard: New dashboard interface with charts, metrics, and data tables
  - Advanced Data Access Layer: Dedicated analytics service with store management and data utilities
  - Enhanced Search Functionality: Improved analytics search with better query handling and routing
  - Comprehensive Testing: Full test coverage for all new components and services
  - User Agent Parsing: Browser detection and analysis utilities
  - Cube Query Builder: Advanced query construction for analytics data
  - Error Handling: Dedicated error component for analytics-related issues

##  Technical Changes

###  New Libraries and Components

  - Data Access Library (libs/portlets/dot-analytics/data-access/):
    - Analytics service with comprehensive API integration
    - Dashboard store with coordinated state management
    - Type definitions for entities, cube queries, and common types
    - Utilities for browser parsing, cube queries, and data processing
  - Dashboard Components:
    - Chart visualization component with configurable displays
    - Metrics component for key performance indicators
    - Filters component for data refinement
    - Table component for detailed data views
    - Error handling component

  ### Enhanced Features

  - Routing Integration: Analytics routes properly integrated into main app routing
  - Health Guard: Analytics health check guard for route protection
  - Improved Search: Refactored search functionality with better performance
  - Localization: Updated language properties for analytics features

###  File Changes Summary

  - 63 files changed: 6,521 additions, 539 deletions
  - New libraries: Complete data access layer with services, stores, and utilities
  - Enhanced components: Dashboard, search, and error handling improvements
  - Configuration updates: Routing, project configuration, and localization

  ### Test Coverage

  - Comprehensive unit tests for all new components and services
  - 397+ test cases for analytics service
  - Extensive testing for dashboard components, utilities, and stores
  - Mock implementations for external dependencies

###  Test Plan

  - Verify analytics dashboard loads correctly
  - Test chart visualization with different data sets
  - Validate filtering and search functionality
  - Confirm error handling for network issues
  - Test responsive design across different screen sizes
  - Verify proper routing and navigation
  - Test health guard functionality

### Screenshots

https://github.com/user-attachments/assets/50d75422-39ff-47be-8bcc-2126d60ef238



This PR fixes: #32724

oidacra committed:
feat(analytics): Add data access library for analytics dashboard

- Introduced a new  library for the analytics dashboard, including:
  - TypeScript configuration and ESLint setup.
  - Core service () for fetching dashboard data.
  - State management using NgRx signals for dashboard state.
  - Mock data and constants for testing and development.
  - Components for displaying metrics, charts, and tables in the dashboard.
- Updated the main analytics portlet to integrate the new data access library and adjusted routing accordingly.

- Enhances modularity and separation of concerns in the analytics feature set.
- Simplifies data fetching and state management for the dashboard.
- Provides a structured approach to building and testing analytics components.

oidacra committed:
feat(analytics): Enhance analytics dashboard with new features and optimizations

### Summary of Changes
- Updated  to include a new redirect for .
- Expanded the data access library with new types and utility functions for analytics data processing.
- Introduced a  utility for comprehensive browser and device detection.
- Enhanced the  with additional methods for fetching top page performance data.
- Improved the  component with loading states and error handling.
- Refactored the dashboard components to utilize new styles and structure for better maintainability.
- Added constants for time period options and table configurations to streamline data handling.

### Benefits
- Provides a more robust and user-friendly analytics dashboard experience.
- Improves data fetching efficiency and error management.
- Enhances code organization and maintainability through modular design.

oidacra committed:
feat(analytics): Add comprehensive tests and enhance dashboard components

oidacra committed:
feat(analytics): Enhance dashboard functionality with coordinated data loading and improved state management

zJaaal commented on core-web/libs/portlets/dot-analytics/data-access/src/lib/types/cubequery.types.ts:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Field types for the query builder
+ */
+export type DimensionField = 'path' | 'pageTitle' | 'userAgent' | 'createdAt' | 'eventType';
+
+export type MeasureField = 'totalRequest' | 'totalSessions' | 'totalUser';
+
+export type FilterField = 'eventType' | 'path' | 'userAgent';
+
+export type OrderField = 'totalRequest' | 'totalSessions' | 'createdAt' | 'path' | 'pageTitle';
```
Same question here

zJaaal commented on core-web/libs/portlets/dot-analytics/data-access/src/lib/utils/browser/userAgentParser.ts:
```diff
@@ -0,0 +1,242 @@
+export interface ParsedUserAgent {
+    browser: {
+        name: string;
+        version: string;
+    };
+    device: {
+        type: 'mobile' | 'tablet' | 'desktop';
+        vendor?: string;
+        model?: string;
+    };
+    os: {
+        name: string;
+        version?: string;
+    };
```
Let's create a type for every property, to prevent issues in the future and better mocking

zJaaal commented on core-web/libs/portlets/dot-analytics/data-access/src/lib/utils/data/analytics-data.utils.ts:
```diff
@@ -0,0 +1,242 @@
+ */
+export interface ChartData {
+    labels: string[];
+    datasets: {
+        label: string;
+        data: number[];
+        borderColor?: string;
+        backgroundColor?: string | string[];
+        borderWidth?: number;
+        fill?: boolean;
+        tension?: number;
+    }[];
```
let's create a type for this to enhance readablitity

zJaaal commented on core-web/libs/portlets/dot-analytics/portlet/src/lib/dot-analytics-dashboard/components/dot-analytics-dashboard-chart/dot-analytics-dashboard-chart.component.scss:
```diff
@@ -0,0 +1,61 @@
+
+.chart-card {
+    ::ng-deep .p-card-body {
+        padding: 1rem;
```
use spacing variables

zJaaal commented on core-web/libs/portlets/dot-analytics/portlet/src/lib/dot-analytics-dashboard/components/dot-analytics-dashboard-chart/dot-analytics-dashboard-chart.component.scss:
```diff
@@ -0,0 +1,61 @@
+
+    &--pie {
+        ::ng-deep .p-skeleton {
+            width: 200px;
+            height: 200px;
```
Use rem values

zJaaal commented on core-web/libs/portlets/dot-analytics/portlet/src/lib/dot-analytics-dashboard/components/dot-analytics-dashboard-metrics/dot-analytics-dashboard-metrics.component.scss:
```diff
@@ -0,0 +1,51 @@
+
+.metric-card {
+    ::ng-deep .p-card-body {
+        padding: 1rem;
```
use spacing variables

zJaaal commented on core-web/libs/portlets/dot-analytics/portlet/src/lib/dot-analytics-dashboard/components/dot-analytics-dashboard-table/dot-analytics-dashboard-table.component.html:
```diff
@@ -0,0 +1,106 @@
+
+            <!-- Skeleton Pagination -->
+            <div class="flex justify-content-center align-items-center p-3">
+                <p-skeleton height="1.5rem" width="250px"></p-skeleton>
```
should we use rem for width too?

zJaaal commented on core-web/libs/portlets/dot-analytics/portlet/src/lib/dot-analytics-dashboard/constants/dot-analytics.constants.ts:
```diff
@@ -0,0 +1,52 @@
+];
+
+/** Default time period selection (1 week) */
+export const DEFAULT_TIME_PERIOD: TimeRange = 'from 7 days ago to now';
```
Oh, you have the constant here, check if you can reuse it or go for the enum approach

zJaaal commented on core-web/libs/portlets/dot-analytics/portlet/src/lib/dot-analytics-dashboard/dot-analytics-dashboard.component.scss:
Check if we can change the px values for rem ones

oidacra committed:
feat(analytics): Refactor and enhance analytics dashboard components

nicobytes commented on core-web/libs/portlets/dot-analytics/data-access/jest.config.ts:
```diff
@@ -0,0 +1,22 @@
+/* eslint-disable */
+export default {
+    displayName: 'analytics-data-access',
```
@oidacra I suggest using the prefix:  "portlet" like `portlets-dot-analytics` In this case `portlets-dot-analytics-data-access` yeah, so long, but to follow the pattern

nicobytes commented on core-web/libs/portlets/dot-analytics/data-access/src/lib/services/dot-analytics.service.ts:
```diff
@@ -0,0 +1,148 @@
+     * Total de pageviews en el período especificado
+     */
+    totalPageViews(
+        timeRange: TimeRange = 'from 7 days ago to now'
```
Maybe we could use `TimeRange.LAST_7_DAYS`

oidacra replied:
I did this better
```
/**
 * Default time range for analytics queries.
 */
export const DEFAULT_TIME_RANGE = TimeRange.LAST_7_DAYS;
```

instead of exposing the const.

nicobytes commented on core-web/libs/portlets/dot-analytics/data-access/src/lib/types/common.types.ts:
```diff
@@ -0,0 +1,27 @@
+    LAST_7_DAYS: 'from 7 days ago to now',
+    LAST_30_DAYS: 'from 30 days ago to now',
+    THIS_WEEK: 'this week'
+} as const;
```
🥇

nicobytes commented on core-web/libs/portlets/dot-analytics/portlet/project.json:
```diff
@@ -1,5 +1,5 @@
 {
-    "name": "portlets-dot-analytics",
+    "name": "analytics-portlets",
```
Other portlets follow the prefix `portlets-[my-portlet]`:

<img width="434" height="189" alt="Image" src="https://github.com/user-attachments/assets/1f1f49c1-f216-4b8f-a0d0-227673335c56" />

nicobytes commented on core-web/libs/portlets/dot-analytics/portlet/src/lib/dot-analytics-dashboard/components/dot-analytics-dashboard-chart/dot-analytics-dashboard-chart.component.scss:
```diff
@@ -0,0 +1,125 @@
+@use "variables" as *;
+
+// Local variables for chart component
+$chart-pie-size: 200px;
```
Should we use rem here?

nicobytes commented on core-web/libs/portlets/dot-analytics/portlet/src/lib/dot-analytics-dashboard/components/dot-analytics-dashboard-metrics/dot-analytics-dashboard-metrics.component.scss:
```diff
@@ -0,0 +1,55 @@
+.metric-card {
+    ::ng-deep .p-card-body {
+        padding: $spacing-3;
+        min-height: 114px;
```
use rem

oidacra commented on core-web/libs/portlets/dot-analytics/portlet/src/lib/dot-analytics-dashboard/dot-analytics-dashboard.component.scss:
```diff
@@ -0,0 +1,111 @@
+
+    // Element: metrics
+    &__metrics {
+        display: flex;
```
fixed this and others.

oidacra committed:
feat(analytics): Update analytics data access library and dashboard components

oidacra committed:
feat(analytics): Refactor analytics search components and improve routing

### Summary of Changes
- Removed unused imports and simplified the .
- Added a new error route and integrated  for route protection in the analytics portlet.
- Refactored the  component to enhance the layout and improve user interaction.
- Updated the analytics search store to remove unnecessary dependencies and streamline state management.

oidacra committed:
feat(analytics): Update analytics search tests and improve query examples

### Summary of Changes
- Refactored the test suite for the DotAnalyticsSearchComponent to enhance readability and maintainability.
- Added new test cases for rendering components and handling query execution.
- Updated the analytics query examples in the utils file to include additional scenarios and improve clarity.
- Reorganized the Language.properties file to ensure consistency in analytics search messages.

oidacra committed:
fix(analytics): Refactor JSON parsing and improve code formatting in utils and store

oidacra committed:
fix(analytics): Update project name in JSON configuration for consistency

oidacra committed:
fix(tests): Update expected portlet count in SerializationHelperTest to reflect changes in portlet.xml

oidacra committed:
fix(analytics): Refactor UserAgentParser to handle user agent input more flexibly

oidacra committed:
feat(analytics): Enhance pageViewTimeLine granularity handling and add unit tests
